working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+```dart
+// dart:concurrent
+
+class Lock implements Shareable {
+  void acquireSync();
+  bool tryAcquireSync({Duration? timeout});
+
+  void release();
+  
+  Future<void> acquire();
+  Future<bool> tryAcquire({Duration? timeout});
+}
+
+class Condition implements Shareable {
+  bool waitSync(Lock lock, {Duration? timeout});
+  Future<bool> wait(Lock lock, {Duration? timeout});
+  
+  void notify();
+  void notifyAll();
+}
+```
+
+# Appendix 
+
+## Memory Models
+
+Memory model describes the range of possible behaviors of multi-threaded programs which read and write shared memory. Programmer looks at the memory model to understand how their program will behave. Compiler engineer looks at the memory model to figure out which code transformations and optimization are valid. The table below provides an overview of memory models for some widely used languages. 
+
+| Language   | Memory Model                                                 |
+| ---------- | ------------------------------------------------------------ |
+| C#         | Language specification itself (ECMA-334) does not describe any memory mode. Instead the memory model is given in Common Language Infrastructure (ECMA-335, section *I.12.6 Memory model and optimizations*). ECMA-335 memory model is relatively weak and CLR provides stronger guarantees documented [here](https://github.com/dotnet/runtime/blob/main/docs/design/specs/Memory-model.md). See [dotnet/runtime#63474](https://github.com/dotnet/runtime/issues/63474) and [dotnet/runtime#75790](https://github.com/dotnet/runtime/pull/75790) for some additional context. |
+| JavaScript | Memory model is documented in [ECMA-262 section 13.0 *Memory Model*](https://262.ecma-international.org/13.0/#sec-memory-model)*.* This memory model is fairly straightforward: it guarantees sequential consistency for atomic operations, while leaving other operations unordered. |
+| Java       | Given in [Java Language Specification (JLS) section 17.4](https://docs.oracle.com/javase/specs/jls/se19/html/jls-17.html#jls-17.4) |
+| Kotlin     | Not documented. Kotlin/JVM effectively inherits Java's memory model. Kotlin/Native - does not have a specified memory model, but likely follows JVM's one as well. |
+| C++        | Given in the section [Multi-threaded executions and data races](https://eel.is/c++draft/intro.multithread) of the standard (since C++11). Notably very fine grained |
+| Rust       | No official memory model according to [reference](https://doc.rust-lang.org/reference/memory-model.html), however it is documented to "[blatantly inherit C++20 memory model](https://doc.rust-lang.org/nomicon/atomics.html)" |
+| Swift      | Defined to be consistent with C/C++. See [SE-0282](https://github.com/apple/swift-evolution/blob/main/proposals/0282-atomics.md) |
+| Go         | Given [here](https://go.dev/ref/mem): race free programs have sequential consistency, programs with races still have some non-deterministic but well-defined behavior. |
+
```

Comment:
Can we explicitly state how what we propose here relates to that?

Reply:
Added a comment here. tldr is that I think it is too early for us to decide our memory model - but we should probably use JS one.